#159204691: user should receive a json web token on authentication

chidioguejiofor opened the pull request:
#### What does this PR do?
- adds  signup test to the app
- adds a service folder to the app
- creates a .env-sample file to the app

#### Description of Task to be completed?
- as a user, I should receive a JSON Web Token on authentication

#### How should this be manually tested?
- run the command `npm install` to install all the dependencies
- run `npm start` in order to start the server
- create a post request to **localhost:3000/api/users/**. The body of this request should be similar to 
have the properties in the JSON below:
{
	"user" : {
		"email": "chidi@email.com",
		"username": "username101",
		"password": "password1"
	}
}
- you should receive a response like this:

{
    "data": {
        "email": "chidi@email.com",
        "username": "username101",
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImNoaWRpQGVtYWlsLmNvbSIsInVzZXJuYW1lIjoidXNlcm5hbWUxMDEiLCJwYXNzd29yZCI6InBhc3N3b3JkMSIsImlhdCI6MTUzMjQ1NDE0NCwiZXhwIjoxNTMyNzEzMzQ0fQ.njMCExUNOhkQwLk4vzE9ktCQvG1LBaWqNXfCnpprumw"
    }
}


#### Any background context you want to provide?
- Ensure that mongodb is set up and running
#### What are the relevant pivotal tracker stories?
- 159204691
#### Screenshots (if appropriate):
<img width="1146" alt="screen shot 2018-07-24 at 6 45 28 pm" src="https://user-images.githubusercontent.com/38565349/43156353-cdd1683a-8f71-11e8-88f5-ccd461815508.png">

#### Questions:
- None

rhenshaw56 commented on .env-sample:
```diff
@@ -0,0 +1 @@
+JSON_SECRET=anysecretgoeshere
```
an appropriate name would be `AUTH_SECRET` or `JWT_SECRET` cos `JSON_SECRET` sounds abit confusing, don't you think ?

chidioguejiofor replied:
Thanks

rhenshaw56 commented on routes/api/users.js:
```diff
@@ -79,7 +81,17 @@ router.post("/users", function(req, res, next) {
     user.save()
         .then(function() {
-            return res.json({ user: user.toAuthJSON() });
+            console.log(TokenService);
```
get rid of the `console` statements

rhenshaw56 commented on routes/api/users.js:
```diff
@@ -79,7 +81,17 @@ router.post("/users", function(req, res, next) {
         .then(function() {
-            return res.json({ user: user.toAuthJSON() });
+            console.log(TokenService);
+            TokenService.createToken(req.body.user, '3d', function(token){
```
first off! transform this to either return just the token and not make use of a callback cos then you're now forced to return a response in the context of the `TokenService.createToken`. What if you still had to do something else with the user data.

Also, define the token expiry inside where you're using it, cos it's redundant if you have to pass in a default expiry value to the function, and that just makes the function impure.

Finally, I'd like to hope that since you're creating the token with the whole `req.body.user` that the user's password ain't in there somehow

rhenshaw56 commented on routes/api/users.js:
```diff
@@ -2,6 +2,8 @@ const mongoose = require("mongoose");
 const router = require("express").Router();
 const passport = require("passport");
 const User = mongoose.model("User");
+const TokenService = require('../../services/TokenService');
```
this pascal case naming convention should only hold if `TokenService` is a class

chidioguejiofor replied:
Okay noted. I would see how to convert this to a class

chidioguejiofor replied:
I think I could convert this to a class once the refactor to es6 is done.

rhenshaw56 commented on test/signup.test.js:
```diff
@@ -0,0 +1,69 @@
+                request.post('/api/users/')
+                .send({
+                    user: {
+                        email: 'goodemailformat@goodemail.com',
```
there's this really nice npm module for having mock data called [faker](https://www.npmjs.com/package/faker) you could use that instead

rhenshaw56 commented on test/signup.test.js:
```diff
@@ -0,0 +1,69 @@
+                });
+            });
+        })
+     });
```
don't you think it'd be nice to test when a user tries to signup without inputing any details or some invalid details or a user tries to signup but they were already members. What do you think ?

chidioguejiofor replied:
I think that would be nice

chidioguejiofor replied:
I would add that

rhenshaw56 commented:
I see you integrated testing setup in these. could you quickly create a separate pr that shows the test setup so I could quickly get that in an then you can rebase this and update the tests

megame24 commented on test/signup.test.js:
```diff
@@ -0,0 +1,69 @@
+                        password: 'passthat1password',
+                    }
+                })
+                .expect(201, done);
```
I suggest you wrap your assertions with a `.end()` block, which should be independent of your `done` function this will ensure tests that deal with promises terminate properly.

megame24 commented on services/TokenService.js:
```diff
@@ -0,0 +1,11 @@
+
+function createToken(payload,expiresIn, callback){
+    jwt.sign(payload,process.env.JSON_SECRET, {expiresIn}, (err, token)=> {
+        callback(token);
```
I believe we should stick to promises in place of callbacks to ensure consistency. Also in this use case, you should just return the token without a callback.

rhenshaw56 commented on config/passport.js:
```diff
@@ -1,7 +1,5 @@
 const passport = require("passport");
 const LocalStrategy = require("passport-local").Strategy;
```
change LocalStategy to camelCase

chidioguejiofor replied:
`LocalStrategy` is actually a class. 
If you look downwards it was instantiated when `passport.use` function was called

rhenshaw56 commented on services/TokenService.js:
```diff
@@ -0,0 +1,9 @@
+const jwt = require('jsonwebtoken');
+
+function createToken(payload,expiresIn, callback){
+    return jwt.sign(payload,process.env.AUTH_SECRET, {expiresIn});
```
fix indentation

rhenshaw56 commented on test/signup.test.js:
```diff
@@ -0,0 +1,59 @@
+const request = require('supertest')(app);
+const db = require('../models/');
+const mocha = require('mocha');
+const faker = require('faker');
```
nice!

rhenshaw56 commented on test/signup.test.js:
```diff
@@ -0,0 +1,59 @@
+     describe('/api/users/', function(){
+
+        describe('status code', function(){
+            it('should be equal to 201', function(done){
```
a proper wording would be `should be equal to 201 when a user signs up successfully`

chidioguejiofor replied:
Noted

rhenshaw56 commented on test/signup.test.js:
```diff
@@ -0,0 +1,59 @@
+            });
+        })
+        describe('request body', function(){
+            it('should have a data property that contains a token string property', function(done){
```
`should have a token property`

chidioguejiofor replied:
Do you mean that I should not put the token in the data property? What I have now is:
{
   data:{
     username: 'username',
    "token": "eycodac dscds",
    ...
  }
}

chidioguejiofor replied:
should I change it to

{
  data:{
         username: 'username',
         ...
   },
  "token": "eycodac dscds",
   .....
}

rhenshaw56 replied:
change ~should have a data property that contains a token string~ to **should have a token property**

rhenshaw56 commented on test/signup.test.js:
```diff
@@ -0,0 +1,59 @@
+
+describe('POST requests', function() {
+    beforeEach(function(done){
+      db.User.destroy({
```
so since this feature flows through the the api endpoint not the db, a good approach would be hitting the signup endpoint with valid details for signup first and then do the your thing

chidioguejiofor replied:
I don't understand what you mean

rhenshaw56 commented:
you got a whole lot of indentation issues, you got to fix that mehn!

chidioguejiofor commented:
I would correct those indentation issues once linting is set up. I cannot really see all the errors without linting

rhenshaw56 commented on routes/api/users.js:
```diff
@@ -1,6 +1,9 @@
 const User = require("../../models/").User;
+const TokenService = require('../../services/TokenService');
+const dotenv = require('dotenv/config');
+const AuthValidator = require('../../validators/AuthValidator');
```
put the validator in a middlewares folder

chidioguejiofor replied:
I actually synced with my teammates and found that someone else was working on validator so i removed it

rhenshaw56 commented on test/signup.test.js:
```diff
@@ -0,0 +1,127 @@
+function generateUser(){
+    const user = {
+        email: faker.internet.email(),
+        username:faker.name.lastName() + Date.now(),
```
instead if using `Date.now()` can you just trim out the email part from the `@` symbol and use that for the `username`

megame24 commented on index.js:
```diff
@@ -12,6 +12,7 @@ const fs = require("fs"),
 require('dotenv').config();
 
 const isProduction = process.env.NODE_ENV === "production";
+const isTest = process.env.NODE_ENV === "test";
```
is the variable `isTest` necessary, I don't see anywhere you are making use of it.

megame24 commented on test/signup.test.js:
```diff
@@ -0,0 +1,127 @@
+const chai = require('chai');
+const app = require('../index');
+const request = require('supertest')(app);
```
I have set up the tests to use `chai-http`, can you check out the `index.test.js` file and follow the testing pattern used there so that everyone's test will be consistent going forward.

megame24 commented on .travis.yml:
```diff
@@ -1,5 +1,11 @@
+before_script:
+  - npm install -g sequelize-cli
+  - psql -c 'create database AuthorsHavenTest;' -U postgres
+  - sequelize db:migrate --env test
```
This code is not needed, it already exists in the `pretest` script.

megame24 commented on config/passport.js:
```diff
@@ -10,14 +10,29 @@ passport.use(
+                    }
+                    if (!(hash ===  user.hash)) {
+                        return done(null, false, {
+                            errors: { "email": "is invalid" }
```
The error message should be `{ "password": "is invalid" }` for this logic branch.

megame24 commented on models/User.js:
```diff
@@ -8,18 +8,13 @@ module.exports = (sequelize, DataTypes) => {
       required: true,
-      validate: {
-        is: /^[a-zA-Z0-9]+$/i,
-      },
```
Schema validation has been removed, attend to it.

megame24 commented on models/User.js:
```diff
@@ -8,18 +8,13 @@ module.exports = (sequelize, DataTypes) => {
       required: true,
-      validate: {
-        isEmail:true,
-      },
```
Schema validation has been removed, attend to it.

mppdg commented on package.json:
```diff
@@ -6,12 +6,15 @@
     "start": "node index.js",
     "test": "cross-env NODE_ENV=test mocha ./test/**/*.js",
-    "coverage": "cross-env NODE_ENV=test ./node_modules/istanbul/lib/cli.js cover ./node_modules/mocha/bin/_mocha",
-    "lint": "eslint ."
```
Why did you remove the `lint` script from `package.json`?

megame24 commented on package.json:
```diff
@@ -6,12 +6,15 @@
-    "coverage": "cross-env NODE_ENV=test ./node_modules/istanbul/lib/cli.js cover ./node_modules/mocha/bin/_mocha",
-    "lint": "eslint ."
+    "coverage": "npm run pretest && cross-env NODE_ENV=test ./node_modules/istanbul/lib/cli.js cover ./node_modules/mocha/bin/_mocha && npm run posttest",
+    "pretest": "sequelize --env test db:migrate && sequelize --env test db:seed",
```
You should undo migration before running `db:migrate` to ensure the tables are dropped before being recreated in a situation where your test ended with an error, and didn't get the chance to drop the table(`db:migrate:undo:all`). I suggest you edit the `pretest` script to look like this `npm run posttest && sequelize --env test db:migrate && sequelize --env test db:seed`

tutugodfrey commented on package.json:
```diff
@@ -6,12 +6,15 @@
+    "pretest": "sequelize --env test db:migrate && sequelize --env test db:seed",
+    "posttest": "sequelize --env test db:migrate:undo:all",
+    "migration": " sequelize db:migrate:undo:all && sequelize db:migrate   "
   },
```
You have removed the script for linting on line 10. Can you please add it so we can always lint to clear errors before pushing

tutugodfrey commented on config/passport.js:
```diff
@@ -10,14 +10,29 @@ passport.use(
+                where: {
+                    email: email
+                }
+            })
```
@chidioguejiofor can you use object destructing in the email field to avoid hound shouting.

``` 
where: {
  email,
}
```
will be ok

megame24 commented on routes/api/users.js:
```diff
@@ -1,6 +1,8 @@
 const router = require("express").Router();
 const passport = require("passport");
 const User = require("../../models/").User;
+const TokenService = require('../../services/TokenService');
```
Since `TokenService` is not a constructor, change the capital `T` to a small letter `t`.

megame24 commented on test/auth.test.js:
```diff
@@ -0,0 +1,190 @@
+            }); 
+    
+        });
+    });
```
To fix your indentation issues, you can set your editor spaces to `2` at the bottom left corner of your VS code editor. Then press `option` + `shift` + `f` on the file to auto correct the indentation.

mppdg commented on package.json:
```diff
@@ -33,6 +37,7 @@
     "request": "^2.87.0",
     "sequelize": "^4.38.0",
     "slug": "^0.9.1",
+    "supertest": "^3.1.0",
```
You should remove the `supertest` dependency since you don't have the need for it.

tutugodfrey commented on test/auth.test.js:
```diff
@@ -0,0 +1,190 @@
+        
+        describe('loging in with all the required fields', ()=>{
+            it('should return a 200 status code has a data property with a token', (done)=>{
+    
```
Please close all trailing the spaces for cleaner code

tutugodfrey commented on test/auth.test.js:
```diff
@@ -0,0 +1,190 @@
+                .end((err, res)=>{
+                    expect(res.body)
+                       .to.not.have.property('data');
+                    done();
```
For every occurence of
```
.end((err, res)=>{
                    expect(res.body)
                       .to.not.have.property('data');
```
you should also check for the value in `data` object. That way you are sure the value is not the same with what is in

```
                .end((err, res)=>{
                    expect(res.status).to.equal(201 );
                    expect(res.body).to.haveOwnProperty('data');
                    expect(res.body.data).to.haveOwnProperty('token');
                    done();
                })
```

JCanaks commented on test/auth.test.js:
```diff
@@ -0,0 +1,190 @@
+            });
+        });
+       
+        describe('sloging in with no username', ()=>{
```
Correct typo "sloging in" to logging in

JCanaks commented on test/auth.test.js:
```diff
@@ -0,0 +1,190 @@
+const models = require('../models');
+chai.use(chaiHttp);
+
+function generateUser(){
```
This function 'generateUser' is never used in the test. Suggest removing it

JCanaks commented on test/auth.test.js:
```diff
@@ -0,0 +1,170 @@
+        });
+    });
+
+    describe('sloging in with no username', () => {
```
change 'sloging' to 'logging'

rhenshaw56 commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+package-lock.json
```
pls remove package-lock.json from here

rhenshaw56 commented on index.js:
```diff
@@ -1,34 +1,33 @@
 app.use(
   session({
-    secret: "authorshaven",
+    secret: 'authorshaven',
```
put this secret in a .env file

rhenshaw56 commented on index.js:
```diff
@@ -39,25 +38,26 @@ if (!isProduction) {
+  app.use((err, req, res) => {
     // do not log errors on test environment, returning it is good enough
-    if(process.env.NODE_ENV !== 'test') console.log(err.stack);
+    if (process.env.NODE_ENV !== 'test') console.log(err.stack); // eslint-disable-line
```
remove the eslint disabling

rhenshaw56 commented on routes/api/users.js:
```diff
@@ -1,87 +1,105 @@
+        email: user.email,
+      };
 
+      const token = tokenService.createToken(payload, '3d');
```
I think I told you something about not directly passing `'3d'` into this call here, define it where you're using it

rhenshaw56 commented on seeders/20180731154458-User.js:
```diff
@@ -0,0 +1,11 @@
+  up: queryInterface => queryInterface.bulkInsert('Users', [{
+    username: 'usename10101',
+    email: 'emadasfa21il@email.com',
+    password: 'passtheword'
```
read this details from `.env`

Commit:
feat(userSignup): as a user I should receive token on singup

- updates travis.yml
- adds missing tests to the app
- updates tests to use fakers
- changes TokenService.createToken to return a token instead of using callback
- jsonwebtoken npm module to the app
- creates and adds data to .env-sample
- ensures that previously added tests are passing
- adds .env to gitignore
- adds module exports to the .js files
- adds missing tests to the app

[ #159204691]

henshawrowland56 commented on index.js:
```diff
@@ -10,6 +10,7 @@ import {} from './models/User';
 import {} from 'dotenv/config';
```
`import { config } from 'dotenv'`;
`import { passport } from './config'`
